engine/src/main/java/org/hibernate/validator/engine/HibernateConstraintViolation.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.engine;
+
+import javax.validation.ConstraintViolation;
+
+/**
+ * A custom {@link ConstraintViolation} which allows to get additional information for a constraint violation.
+ *
+ * @since 5.3
+ */
+public interface HibernateConstraintViolation<T> extends ConstraintViolation<T> {
+	/**
+	 * @return an object that further describes the violation or {@code null}.
+	 */
+	Object getViolationContext();
```

Comment:
> In the beginning I thought about this too and actually even wanted to go one step further so that the user defines the violation context type on the custom constraint annotation.

The problem is that this would make the use of a custom constraint mandatory and one could use the custom context with a default constraint just using a custom `ConstraintValidator` implementation. I think for now I opt for either plain object or the approach where the expected type is passed to the getter. If no context is set or one of the wrong type `null` is returned. WDYT?